<a href="https://colab.research.google.com/github/rdevia/text_mining_unrn/blob/main/finetunning_1123.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TRABAJO FINAL - TEXT MINING - PARTE B


**Raul Devia**



In [1]:
!pip install datasets seaborn
! pip install -U accelerate
! pip install -U transformers



   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 362.1/362.1 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 43.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 42.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 43.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 20.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 68.8 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalli

1- Crear una cuenta en HUGGINGFACE

https://huggingface.co/

In [2]:
from huggingface_hub import notebook_login
from google.colab import userdata

notebook_login()

In [3]:
pip install -U datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.5/491.5 kB 14.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 19.9 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.2
    Uninstalling fsspec-2025.3.2:
      Successfully uninstalled fsspec-2025.3.2
  Attempting uninstall: datasets
    Found existing installation: datasets 2.14.4
    Uninstalling datasets-2.14.4:
      Successfully uninstalled datasets-2.14.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.2 requires fsspec==2025.3.2, but you have fsspec 2025.3.0 which is incompatible.


In [4]:
from datasets import load_dataset
import pandas as pd
from datetime import datetime, timedelta
from tqdm.auto import tqdm
from collections import Counter

Análisis exploratorio inicial

https://huggingface.co/datasets/piuba-bigdata/contextualized_hate_speech



*   ¿Cómo es este dataset?

*   ¿Que estructura es ds1?¿cuantos conjuntos tiene?






1- carguen el dataset

In [5]:
from datasets import load_dataset

ds1 = load_dataset("piuba-bigdata/contextualized_hate_speech")



/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/3.55k [00:00<?, ?B/s]

dataset_infos.json:   0%|          | 0.00/1.98k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/107M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/1.52M [00:00<?, ?B/s]

dev-00000-of-00001.parquet:   0%|          | 0.00/26.8M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/36420 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/11343 [00:00<?, ? examples/s]

Generating dev split:   0%|          | 0/9106 [00:00<?, ? examples/s]

In [6]:
ds1

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 36420
    })
    test: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 11343
    })
    dev: Dataset({
        features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
        num_rows: 9106
    })
})

El dataset 1 contiene 56869 tw etiquetados.
El dataset se dividió en 36420 de train, test:11343 y vaLidacion:9106

In [7]:
ds1_train=ds1["train"]
ds1_train

Dataset({
    features: ['id', 'title', 'text', 'context_tweet', 'HATEFUL', 'body', 'CALLS', 'WOMEN', 'LGBTI', 'RACISM', 'CLASS', 'POLITICS', 'DISABLED', 'APPEARANCE', 'CRIMINAL'],
    num_rows: 36420
})

1) Dataset de etiquetado

https://huggingface.co/datasets/piuba-bigdata/contextualized_hate_speech

Este es el dataset original de etiquetado. Contiene los siguientes campos:

ID : identificador

TITLE: título de la noticia

TEXT: texto del tweet del comentario al Tweet original

CONTEXT: tweet original

HATEFUL: 1 hateful, 0 not hateful

BODY: texto de la noticia

CALLS: si se realizó o no un llamado a la acción

WOMEN: discriminación hacia mujeres

LGBTI: discriminación hacia grupos LGBTI

RACISM: discriminación asociada a racismo

CLASS: discriminación vinculada a pobreza, villas

POLITICS: odio o violencia vinculada a las ideas políticas

DISABLED: discriminación a las personas con discapacidad

APPEARENCE: discriminación en base al aspecto físico (incluye especialmente la gordofobia)

CRIMINAL: discriminación hacia personas en el sistema penal o similares

**Para el entrenamiento se usaron los siguientes tutoriales y documentos:**

Uso  de base el siguiente tutorial para hacer el finetuning:
https://huggingface.co/blog/sentiment-analysis-python
y este:
https://huggingface.co/docs/transformers/training


La idea es utilizar la librería tranbsformers de huggingface para hacerle finetuning a un BETO original, es decir la version de BERT pre-entrenada con español de GOOGLE pero que no haya sido reentrenado con ningún otro dataset.
Aqui para usar transformers de hugginface
https://github.com/huggingface/transformers?tab=readme-ov-file



In [8]:
import torch
torch.cuda.is_available()

True

### 1. pre-process data

In [9]:
 ds1['train'][0]

{'id': 343726,
 'title': 'Video: salió de la cárcel por el coronavirus y murió de un tiro el mismo día al festejar su libertad',
 'text': '@usuario Uno menos',
 'context_tweet': 'Video: salió de la cárcel por el coronavirus y murió de un tiro el mismo día al festejar su libertad https://t.co/UrJGRAdEvD',
 'HATEFUL': 1,
 'body': 'Un hombre de 46 años, que cumplía una condena en prisión y fue puesto en libertad por el coronavirus, murió el mismo día que dejó la cárcel baleado por un primo, durante los "festejos" por dejar la prisión.\n\nSari Salem Wardat, de 46 años, recibió un disparo desde atrás en un aparente accidente mientras caminaba hacia miembros de la familia y vecinos que se habían reunido para darle la bienvenida a su casa, en Jordania.\n\nUn video que circuló en las redes sociales \u200bmuestra a Sari llegando a su casa en un auto azul rodeado de tiros de "celebración". El primo se acerca de inmediato a Sari, que sostiene una pistola y la apunta hacia el cielo.\n\nEl detenido

In [10]:
##ME quedo solo con las columnas que necesito para la clasificación binaria
ds_train_small= ds1['train'].select_columns(["text","HATEFUL"])
ds_test_small= ds1['test'].select_columns(["text","HATEFUL"])
ds_val_small= ds1['dev'].select_columns(["text","HATEFUL"])

Nos quedamos con las columnas Text y Hateful y se renombra esta última ya que el modelo requiere que se llame "label"

In [11]:
##aca les coloco el mismo nombre que espera el modelo (me parece cualquiera pero no logre que ande sino)
ds_train_small=ds_train_small.rename_column("HATEFUL", "label")
ds_test_small=ds_test_small.rename_column("HATEFUL", "label")
ds_val_small=ds_val_small.rename_column("HATEFUL", "label")

In [12]:
ds_train_small

Dataset({
    features: ['text', 'label'],
    num_rows: 36420
})

In [ ]:
ds_test_small

Dataset({
    features: ['text', 'label'],
    num_rows: 11343
})

In [13]:
# Tomo una muestra de 50 mil para que sea más rapido el entrenamiento
#  uso una semilla, para que pueda reproducirlo, es decir para que mi muestra siempre sea la misma.
# cada uno elija su numero

import random

# Set the random seed for reproducibility
random.seed(17)

# Sample 50,000 random examples from ds_test_small
indices_to_sample = random.sample(range(len(ds_test_small)), 10000)
ds_train_small = ds_train_small.select(indices_to_sample)

ds_train_small


Dataset({
    features: ['text', 'label'],
    num_rows: 10000
})

In [14]:
ds_test_small[0]

{'text': '@usuario Ojalá se funda y cague de hambre', 'label': 0}

In [15]:
!pip install --upgrade jax jaxlib

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 41.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.1/89.1 MB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 100.0 MB/s eta 0:00:00
  Attempting uninstall: ml_dtypes
    Found existing installation: ml-dtypes 0.4.1
    Uninstalling ml-dtypes-0.4.1:
      Successfully uninstalled ml-dtypes-0.4.1
  Attempting uninstall: jaxlib
    Found existing installation: jaxlib 0.5.1
    Uninstalling jaxlib-0.5.1:
      Successfully uninstalled jaxlib-0.5.1
  Attempting uninstall: jax
    Found existing installation: jax 0.5.2
    Uninstalling jax-0.5.2:
      Successfully uninstalled jax-0.5.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.18.0 requires ml-dtypes<0.5.0,>=0.4.0, but you have ml-dtypes 0.5.1 which is incompatible.


In [16]:
from transformers import BertForMaskedLM, BertTokenizer

/usr/local/lib/python3.11/dist-packages/jaxlib/plugin_support.py:71: RuntimeWarning: JAX plugin jax_cuda12_plugin version 0.5.1 is installed, but it is not compatible with the installed jaxlib version 0.6.1, so it will not be used.
  warnings.warn(


Utiizo las librerías de transformers para preparar los datos de entrenamiento

Lo primero es tokenizar en función del modelo seleccionado, en nuestro caso BETO

In [17]:
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("FacebookAI/roberta-large-mnli")

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/688 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Le aplico la función tokenizer del modelo BETO a mis conjuntos, el de entrenamiento, el de test y el de validación

In [18]:
def preprocess_function(examples):
   return tokenizer(examples["text"], truncation=True)

tokenized_train = ds_train_small.map(preprocess_function, batched=True)
tokenized_test = ds_test_small.map(preprocess_function, batched=True)
tokenized_val = ds_val_small.map(preprocess_function, batched=True)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

Map:   0%|          | 0/11343 [00:00<?, ? examples/s]

Map:   0%|          | 0/9106 [00:00<?, ? examples/s]

In [19]:
tokenized_train[0]

{'text': '@usuario acaso será el cerdo de bolsonaro?',
 'label': 1,
 'input_ids': [0,
  1039,
  687,
  257,
  7718,
  4285,
  281,
  139,
  6821,
  1526,
  1615,
  25380,
  5016,
  263,
  24724,
  1478,
  5191,
  116,
  2],
 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

Prueben varios casos

In [20]:
tokenized_train[2100]

{'text': '@usuario Reclamale a los que votaste jaa',
 'label': 0,
 'input_ids': [0,
  1039,
  687,
  257,
  7718,
  1223,
  3998,
  424,
  1627,
  10,
  3774,
  1192,
  19314,
  14631,
  1236,
  6621,
  2],
 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

La siguiente función transforma los datos de entrenamiento agregandole el padding

Se usará luego en el Trainer

In [21]:
from transformers import DataCollatorWithPadding
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

## 2 Entreno el modelo.

Voy a entrenar el modelo con el TRAINER de transformers. Instancio el modelo pre-entrenado BETO y aca selecciono 2 labels ya que se hace una clasificación binaria

Modelo: FacebookAI/roberta-large-mnli

Ustedes tendrán que probar otros modelos


In [24]:
from transformers import AutoModelForSequenceClassification
# Add ignore_mismatched_sizes=True to handle the size mismatch in the classification head
model = AutoModelForSequenceClassification.from_pretrained("FacebookAI/roberta-large-mnli", num_labels=2, ignore_mismatched_sizes=True)

Some weights of the model checkpoint at FacebookAI/roberta-large-mnli were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at FacebookAI/roberta-large-mnli and are newly initialized because the shapes did not match:
- classifier.out_proj.bias: found shape torch.Size([3]) in the checkpoint and torch.Size([2]) in the model instanti

In [25]:
#esto es opcional
#import wandb
#wandb.init(mode="disabled")

In [26]:

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score,roc_auc_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

def compute_metrics(eval_pred):

    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    metrics = {}

    metrics['Accuracy'] = accuracy_score(labels, predictions)
    metrics['Precision'] = precision_score(labels, predictions)
    metrics['Recall'] = recall_score(labels, predictions)
    metrics['f1'] = f1_score(labels, predictions)
    metrics['AUC'] = roc_auc_score(labels, predictions)

    return(metrics)



Revision de parametros de TRAINER
https://huggingface.co/docs/transformers/main_classes/trainer




In [27]:
from transformers import TrainingArguments, Trainer

training_args = TrainingArguments(
    output_dir="beto_ejemplo_clase",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,
    weight_decay=0.01,
    eval_strategy="epoch",
    save_strategy="epoch",
    report_to="none",
    load_best_model_at_end=True,
    push_to_hub=True,
)

In [28]:

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_test,
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)



In [29]:
# Train pre-trained model
trainer.train()

TrainOutput(global_step=1250, training_loss=0.3543704345703125, metrics={'train_runtime': 1889.0887, 'train_samples_per_second': 10.587, 'train_steps_per_second': 0.662, 'total_flos': 3279204361855296.0, 'train_loss': 0.3543704345703125, 'epoch': 2.0})

¿Cuanto tiempo les tardó el modelo? el modelo tardó 31 minutos.


¿CUantos EPOCH hizo? 2



¿Qué creen que es un epoch?
un epoch se refiere a un ciclo completo a través de todo el conjunto de datos de entrenamiento.

In [30]:
trainer.evaluate()

{'eval_loss': 0.35015037655830383,
 'eval_Accuracy': 0.8726968174204355,
 'eval_Precision': 0.6954595791805094,
 'eval_Recall': 0.3494713411240957,
 'eval_f1': 0.4651851851851852,
 'eval_AUC': 0.6603317317395043,
 'eval_runtime': 178.0537,
 'eval_samples_per_second': 63.706,
 'eval_steps_per_second': 3.982,
 'epoch': 2.0}

Esta fue la evaluación del entrenamiento que realizamos en clase.



{'eval_loss': 0.30518776178359985,
 'eval_Accuracy': 0.8914749184519086,
 'eval_Precision': 0.7905544147843943,
 'eval_Recall': 0.4284919309961046,
 'eval_f1': 0.555756044749188,
 'eval_AUC': 0.7035608617896928,
 'eval_runtime': 36.244,
 'eval_samples_per_second': 312.962,
 'eval_steps_per_second': 19.562,
 'epoch': 2.0}

Análisis Comparativo y Conclusiones:

Pérdida (eval_loss): La pérdida es menor en la Evaluación 2 (0.305) que en la Evaluación 1 (0.350). Una menor pérdida indica que el modelo se ajustó mejor a los datos de evaluación en la Evaluación 2.
Precisión General (eval_Accuracy): La precisión general es mayor en la Evaluación 2 (0.891) que en la Evaluación 1 (0.872). Esto significa que el modelo de la Evaluación 2 clasificó correctamente un mayor porcentaje de los ejemplos totales.
Precisión (eval_Precision): La precisión es significativamente mayor en la Evaluación 2 (0.790) que en la Evaluación 1 (0.695). Esto indica que, de las predicciones positivas realizadas, un mayor porcentaje fueron correctas en la Evaluación 2. El modelo es más fiable cuando predice que algo es positivo.
Recall (eval_Recall): El recall también es mayor en la Evaluación 2 (0.428) que en la Evaluación 1 (0.349). Si bien el recall sigue siendo relativamente bajo en ambas evaluaciones, la Evaluación 2 logró identificar una mayor proporción de los verdaderos casos positivos.
F1-Score (eval_f1): El F1-score es considerablemente mayor en la Evaluación 2 (0.555) que en la Evaluación 1 (0.465). Dado que el F1-score considera tanto la precisión como el recall, esta mejora sugiere que el modelo de la Evaluación 2 ofrece un mejor equilibrio general entre estas dos métricas.
AUC (eval_AUC): El AUC es mayor en la Evaluación 2 (0.703) que en la Evaluación 1 (0.660). Esto indica que el modelo de la Evaluación 2 tiene una mejor capacidad para distinguir entre las clases positiva y negativa en general.
Tiempo de Ejecución: La Evaluación 2 se completó mucho más rápido (36.244 s) que la Evaluación 1 (178.0537 s). También procesó significativamente más ejemplos por segundo y dio más pasos por segundo. Esto sugiere una mayor eficiencia en el proceso de evaluación, posiblemente debido a un entorno de ejecución diferente o a optimizaciones en el modelo o el código de evaluación.
Conclusiones Generales:

Basándonos en todas las métricas clave (loss, accuracy, precision, recall, F1-score, AUC), la Evaluación 2 muestra un rendimiento superior en comparación con la Evaluación 1.

El modelo en la Evaluación 2 es más preciso, tiene un mayor recall, un mejor equilibrio entre precisión y recall (mayor F1-score) y una mejor capacidad general de distinción entre clases (mayor AUC).
Además, la evaluación fue mucho más rápida en la Evaluación 2, lo que es una ventaja significativa.
En resumen, el modelo que generó la Evaluación 2 parece ser un modelo más robusto y eficiente para la tarea de clasificación




In [31]:
from transformers import pipelines

In [32]:
examples = ["hay que matarlos a todos", "Vamos a la plaza a marchar","Los chinos son todos sucios","Culpa de los bolitas"]


In [33]:
from transformers import pipeline # Import the pipeline function
examples = ["hay que matarlos a todos", "Vamos a la plaza a marchar","Los chinos son todos sucios","Culpa de los bolitas"]
# Specify the 'text-classification' task explicitly
specific_model = pipeline(task="text-classification", model="natdebandi/beto_ejemplo_clase7")
specific_model(examples)

config.json:   0%|          | 0.00/710 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/439M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/248k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/735k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/695 [00:00<?, ?B/s]

Device set to use cuda:0


[{'label': 'LABEL_1', 'score': 0.9847145676612854},
 {'label': 'LABEL_0', 'score': 0.9297218918800354},
 {'label': 'LABEL_1', 'score': 0.977299690246582},
 {'label': 'LABEL_1', 'score': 0.9890073537826538}]